### ISSUE TITLE
Adding an IPv4 route via an IPv6 next hop

### ISSUE THREAD

**Issue by cooperlees** [svinota/pyroute2#894]
Hi all,

This is probbaly my lack of netlink skills, but is there a way I can add an IPv4 route with a IPv6 next hop. This has been around since ~5.2 Linux kernel. To add it via the `ip` command it would look like:

```console
/sbin/ip -4 replace default via inet6 fe80::1337 src 10.69.69.69 table 69 dev eth0
```

I've tried:
```
    with IPRoute() as ipr:
        ipr.route(
            "add",
            prefsrc="10.127.30.4",
            dst="8.8.8.8/32",
            # gateway="fe80::1337",
            gateway="2000:1337:1c:6707::a",
            oif="eth0",
        )
```
- With both link local and global IPv6 addresses
  - Is there a way to make this work today? If not, can we add support. If so, how would we go about that and I am happy to take a stab at it.

Today it results in a
```
OSError: illegal IP address string passed to inet_pton
```

**Comment by svinota** [comment]
Got it. Fixing.

**Comment by cooperlees** [comment]
> Got it. Fixing.

Awesome. Let me know the branch if you like and I'm happy to test ... But if you have a dual stacked host you should just be able to add/replace a route to test it ..

**Comment by svinota** [comment]
… but it seems that the only thing that I have to fix here is the documentation.

So, in your case a little trick is needed:
```from pyroute2 import IPRoute

# 1. instead of RTA_GATEWAY, RTA_VIA should be used in this case
# 2. RTA_OIF should be the interface index as int, you can get it with link_lookup()

with IPRoute() as ipr:
        ipr.route(
            "add",
            prefsrc="10.69.69.69",
            dst="8.8.8.8/32",
            via={"family": 10, "addr": "fe80::1337"}, 
            oif=ipr.link_lookup(ifname="eth0"), 
            table=100,
        )
```

… and I'm to update the docs.

**Comment by cooperlees** [comment]
> … but it seems that the only thing that I have to fix here is the documentation.
> 
> So, in your case a little trick is needed:
> 
> ```
> 
> # 1. instead of RTA_GATEWAY, RTA_VIA should be used in this case
> # 2. RTA_OIF should be the interface index as int, you can get it with link_lookup()
> 
> with IPRoute() as ipr:
>         ipr.route(
>             "add",
>             prefsrc="10.69.69.69",
>             dst="8.8.8.8/32",
>             via={"family": 10, "addr": "fe80::1337"}, 
>             oif=ipr.link_lookup(ifname="eth0"), 
>             table=100,
>         )
> ```
> 
> … and I'm to update the docs.

Nice! I thought this might be the case.

Would you accept a PR if I added an enum for families so we can use friendly names rather than random integers for family? e.g.
```python
    via={"family": AF.INET6, "addr": "fe80::1337"},
# Instead of
    via={"family": 10, "addr": "fe80::1337"},
```

I feel the former / new example is much more pythonic and readable.

**Comment by svinota** [comment]
oh sure you can use something like

```
from socket import AF_INET6
…
     via={"family": AF_INET6, "addr": "fe80::1337"},
```

that's silly me just using int constants for quick-n-dirty checks :)

**Comment by cooperlees** [comment]
Even better. Ok, Lets close this with the documentation PR. Thanks for being awesome. As you know, I love your library!

**Comment by svinota** [comment]
The docs are updated, and on the web as well.

Thanks :)

**Comment by svinota** [comment]
Forgot to add a functional test for that. Closing in a while after adding the test.

**Comment by cooperlees** [comment]
Is there any interest in making gateway just handle IPv6 next hops under the covers? For example make the following just work:

```python
with IPRoute() as ipr:
    eno1 = ipr.link_lookup(ifname="eno1") 
    ipr.route(
        "add",
        dst="8.8.8.8/32",
        gateway="fe80::1",
        oif=eno1,
    )
```

**Comment by svinota** [comment]
I would propose to simplify the `RTA_VIA` syntax so it would accept string as well, guessing the family from the format:

```python
with IPRoute() as ipr:
    eno1 = ipr.link_lookup(ifname="eno1") 
    ipr.route(
        "add",
        dst="8.8.8.8/32",
        via="fe80::1",  # instead of {"family": AF_INET6, "addr": "fe80::1"}
        oif=eno1,
    )
```

TL;DR

Rationale: on one hand, it looks more convenient to have single `gateway` for all the cases. On another hand, I'm trying to keep the `IPRoute` class a simple RTNL API mapping.

And `gateway` corresponds to `RTA_GATEWAY`, and it may become a bit confusing, since `RTA_VIA` is a completely different NLA with a completely different structure: https://github.com/svinota/pyroute2/blob/e8d866b65e085ad19d19a1332a31402d1a1e3fa6/pyroute2.core/pr2modules/netlink/rtnl/rtmsg.py#L96-L121

In other words — when `gateway` should be interpreted as `RTA_GATEWAY` and when as `RTA_VIA`? Only when `dst` and `gateway` have different families?

Beside of heuristics to choose implicitly between `RTA_GATEWAY` and `RTA_VIA` in order to make a request, there is another issue: the response. The dump looks like that:

```python
>>> ipr.route("dump", table=100)
(
  {
    'family': 2,
    'dst_len': 32,
    'src_len': 0,
    'tos': 0,
    'table': 100,
    'proto': 4,
    'scope': 0,
    'type': 1,
    'flags': 0,
    'attrs': [
      ('RTA_TABLE', 100),
      ('RTA_DST', '8.8.8.8'),
      ('RTA_PREFSRC', '192.168.122.28'),
      ('RTA_VIA', {'family': 10, 'addr': 'fe80::1337'}),
      ('RTA_OIF', 2),
    ],
    'header': {
      'length': 84,
      'type': 24,
      'flags': 2,
      'sequence_number': 255,
      'pid': 59410,
      'error': None,
      'target': 'localhost',
      'stats': Stats(qsize=0, delta=0, delay=0),
    },
    'event': 'RTM_NEWROUTE',
  },
)
```

So there will be another source of confusion, unless we try to translate it back.

Thus I try to follow these approaches in `IPRoute`:
1. explicit better than implicit
2. try to avoid any unnecessary semantics change from the RTNL API

**Comment by cooperlees** [comment]
I also agree with explicit over implicit where it makes sense. Since you have a goal with keeping semantics with the RTNL API, this makes sense. Thanks for explaining.

**Comment by cooperlees** [comment]
For anyone who comes across this, the example and documentation can be found here: https://docs.pyroute2.org/iproute.html#module-pyroute2.iproute.linux
